### ISSUE TITLE
Suggestion: Local dimensions() function

### ISSUE THREAD

**Issue by SolarLune** [GoranM/bdx#320]
Hey, there. So, the Gameobject.dimensions() function returns the bounding box dimensions of the object, which can change if the object rotates. It'd be nice if there were a "dimensionsBase()" function, which would return the dimensions of the object without any orientation or scaling changes. Any ideas? 

I'm not good with matrices and Vectors, so an explanation of how to create this function if it's OK'd would be appreciated.

EDIT: OK, I think I got it. Multiply the vector by the inverse of the orientation matrix, then call absolute() on the vector. Anyway, does anyone like the sound of this local dimensions function?

**Comment by rafcolson** [comment]
Yes, I do. But wouldn't "dimensionsLocal" be a better name? Because we already have "velocityLocal" and so fort.

**Comment by SolarLune** [comment]
True, and I do agree. I guess disregarding scaling isn't really necessary, so just having it be local in terms of orientation would be helpful enough.

**Comment by GoranM** [comment]
Perhaps dimensions() should simply return those of the object oriented bounding box?

I don't think we really need axis aligned, do we?

> Multiply the vector by the inverse of the orientation matrix, then call absolute() on the vector.

Wouldn't that simply rotate the vector? 

Imagine an object with an orientation that describes a rotation of pi/4 (45 degrees) around the Z axis: If `d = dimensions()`, and `d.x == d.y`, `inverse_orientation.mult(d)` would produce a vector v where `v.y == 0`.

PS: In the future, it would be best to justify suggestions with something more concrete than "It'd be nice". Why would it be nice? Why do you need it? Answers to those questions tend to be fairly convincing.

**Comment by SolarLune** [comment]
> Perhaps dimensions() should simply return those of the object oriented bounding box?
I don't think we really need axis aligned, do we?

Not to my knowledge, no. Having dimensions simply return the globally aligned dimensions vector would be fine.

> Wouldn't that simply rotate the vector?

It should rotate it back to "identity", so to speak, so it's the normal version of the matrix again.

> PS: In the future, it would be best to justify suggestions with something more concrete than "It'd be nice". Why would it be nice? Why do you need it? Answers to those questions tend to be fairly convincing.

I didn't think I needed to go into deep detail about this, but okay. It's generally undesirable for dimensions() to return a rotated box as it changes the dimensions of the shape, making processes like positioning something one object-width away basically unreliable.

**Comment by rafcolson** [comment]
>It's generally undesirable for dimensions() to return a rotated box as it changes the dimensions of the shape, making processes like positioning something one object-width away basically unreliable.

I agree. I think this would be bad idea. Implementation could be as simple as storing normalized (inverse scaled) Blender dimensions per model as is done with "origins". Then give each object the vector. Don't forget to add updates for cloneNoChildren and replaceModel.

**Comment by GoranM** [comment]
> It should rotate it back to "identity", so to speak, so it's the normal version of the matrix again.

... I'm not really sure what you mean: When multiplying a matrix by a vector, the result is a vector, not a matrix.

Also, based on the scenario I described in my previous post, I think your algorithm is incorrect.

> I didn't think I needed to go into deep detail about this

Simply explaining why you would generally need something is not "deep detail".

> It's generally undesirable for dimensions() to return a rotated box as it changes the dimensions of the shape, making processes like positioning something one object-width away basically unreliable.

It doesn't return a "rotated box". The dimensions method returns a vector that represents the dimensions of an axis aligned bounding box for the given object.

> Implementation could be as simple as storing normalized (inverse scaled) Blender dimensions per model as is done with "origins".

There should be no need to do anything in the exporter: It should be possible to simply save orientation, then set orientation to identity, recalculate the bounding box, then set orientation to previously saved.

At that point, the dimensions of the bullet bounding box should represent the local dimensions.

**Comment by SolarLune** [comment]
You know, I think I can see the benefit of having the dimensions() function work the way it does currently, even though I'd personally rather it not. I also don't want to spend the time to read up on matrices just to deal with this, so I'm just gonna leave it alone. If someone else wants to take a look at it, feel free to. Otherwise, I'll just close this issue.

**Comment by rafcolson** [comment]
>There should be no need to do anything in the exporter: It should be possible to simply save orientation, then set orientation to identity, recalculate the bounding box, then set orientation to previously saved.

In my personal opinion the physics world should never be modified in order to do calculations, even if it would be for a split millisecond.

That said, the current application of dimensions in BDX doesn't really match up with dimensions in Blender, does it? DimensionsLocal would however. But I think the current Dimensions is also useful.

So I would to add it, but I still think we just should get the base dimensions (without scale) in exporter and then use a getter to multiply with scale ().

**Comment by GoranM** [comment]
> I think I can see the benefit of having the dimensions() function work the way it does currently

What's the benefit?

> I also don't want to spend the time to read up on matrices just to deal with this

There's nothing in this issue that requires deep matrix knowledge ... It seems to me like your difficulties stem from misunderstanding what the dimensions method returns (I clarified that in my previous post).

> In my personal opinion the physics world should never be modified in order to do calculations, even if it would be for a split millisecond.

Why not?

> dimensions in BDX doesn't really match up with dimensions in Blender, does it?

It doesn't: dimensions in blender are independent of object orientation.

> I think the current Dimensions is also useful.

Why is it useful?

PS:

*We should really try to get into the habit of explaining why we think certain features/behaviors are useful/beneficial, because that's the most effective way to confirm those beliefs as true, or to discover that they're false.*

**Comment by SolarLune** [comment]
> What's the benefit?

I don't know, it's something that can be used as is if someone wants to know the dimensions, which can change when the object rotates.

I don't know how to "fix" this issue effectively apart from storing the dimensions vector on mesh change. Is that an adequate solution?

As for the question, why did you implement it the way it is? Was it unintentional to have the vector change according to the orientation of the object?

> There's nothing in this issue that requires deep matrix knowledge ... It seems to me like your difficulties stem from misunderstanding what the dimensions method returns (I clarified that in my previous post).

No, I know what the method returns. I said "rotated box", which was wrong.

**Comment by GoranM** [comment]
>  it's something that can be used as is if someone wants to know the dimensions, which can change when the object rotates.

If we can come up with an uncontrived example where current dimensions behavior would be useful, it could be worthwhile to make an additional method, or to just expose the bullet bounding box.

If we can't, then it's probably best to just make dimensions behave as most would expect (given their blender experience), where dimensions are independent from orientation.

> I don't know how to "fix" this issue effectively apart from storing the dimensions vector on mesh change.

As I said before:  It should be possible to simply save orientation, then set orientation to identity, recalculate the bounding box, then set orientation to previously saved.

At that point, the dimensions of the bullet bounding box should represent the local dimensions.

> Was it unintentional to have the vector change according to the orientation of the object?

Yes.

> No, I know what the method returns. I said "rotated box", which was wrong.

Ok ... So why would you think that multiplying the vector by the inverse of the orientation matrix would work? What was the rationale behind that?

**Comment by rafcolson** [comment]
>Why not?

Firstly, if we cross that line, then where will be the line? Secondly, because it's not practical. As you already noticed, developers may need an object's dimensions *before* adding it to the scene. But, in order to use the physics world to calculate the dimensions the object has to be added to the scene.

For example, apes in trees. I want to generate a bunch of trees with apes in it. The trees will be cloned from five types of trees, all with a different geometry. It will be generated with random sizes to give more detail to the scene. Now, what will be the height of every tree? It will be the orientation by the local scale on the Z-axis by the local dimensions on the Z-axis of that tree. So that will be the position for the apes to sit and throw bananas at you.

You could add each tree as you're generating all properties for the tree, but not everyone would prefer to work like that. I would first want to store all properties into a hashMap *before* compilation, when I've implemented Level Of Detail. This would make loading the game take a bit longer, but during the game the LOD will be updated faster.

I hope that explains it. And I agree that we better explain why we think something's better or not. Sorry for being vague there.

**Comment by GoranM** [comment]
> Firstly, if we cross that line, then where will be the line?

... I don't understand. What line?

> developers may need an object's dimensions before adding it to the scene.

If a developer truly *needs* something, they should be able to explain why.

> I hope that explains it.

I'm sorry, but I still don't understand how anything in your example would require knowing the dimensions of a given object before it's actually added to the scene.

You can add the object, and then position it however you wish, based on dimensions, or any other data.

> Let me rephrase myself here, I "guess" the current Dimensions is also useful. I can't think of anything right now, but there must have been a valid reason for it to get into BDX, or else it wouldn't be in there, would it? So maybe I should ask you, or the person who added it?

My response to SolarLune's question covers this: It wasn't intentional.

**Comment by rafcolson** [comment]
>I'm sorry, but I still don't understand how anything in your example would require knowing the dimensions of a given object before it's actually added to the scene.

I'm sorry that I wasn't clear enough then. For example this would have to be the code needed to update only one of the sections around the player, with your proposed implementation of ```dimensionsLocal```:
```java
public void generateTrees(){
	ArrayList<String> treeNames = new ArrayList<String>(asList("tree0", "tree1", "tree2", "tree3", "tree4"));
	int numTrees = 100;
	int size = 1000;
	
	int sectionRadius = (size / numTrees) * 0.5f;
	int numRowsColumns = (int)Math.sqrt(100);
	for (int x = 0;  x < numRowsColumns; x++){
		for (int y = 0;  y < numRowsColumns; y++){
			Vector3f treePosition = new Vector3f(x, y, 0).plus(Random.vector().mul(sectionRadius));
			Vector3f treeScale = Random.random() * 0.25f + 0.75f;
			GameObject tree = scene.add(Random.choice(treeNames));
			tree.position(treePosition);
			tree.scale(treeScale);
			Vector3f apePosition = new Vector3f(treePosition);
			apePosition.z = tree.dimensionsLocal().z * treeScale.z;
			GameObject ape = scene.add("ape");
			ape.position(apePosition);
		}
	}
}
```
But instead it just could be this, that is, if ```dimensionLocal``` wouldn't require adding objects to the scene first:
```java
// first generate data for trees at initialization (or while loading the game)
public void spawnTrees(int sectionIndex){
	HashMap trees = sections.get(sectionIndex).get("trees");
	for (Map.Entry<String, Object> data : trees.entrySet()){
		GameObject tree = scene.add(data.getKey());
		Vector3f[] properties = data.getValue();
		tree.position(properties[0]);
		tree.scale(properties[1]);
		GameObject ape = scene.add("ape");
		ape.position(properties[2]);
	}
}
```
So would it be to much for the programmer to ask to be able to have everything calculated at loading time? This is only a quick example, but you should see the drawback of your proposed implementation. As a programmer you'll be stuck to the limitation it holds.

Most programmers would naturally assume that you don't have to add any objects to the scene in order for something like ```dimensionsLocal``` to work. It sure would be an exception to the rule. Which other method in BDX needs this? Because of that I also think it would be confusing. You would have to add notes about this in the API to counter this, I'm sure.

>... I don't understand. What line?

That's what I meant with crossing the line. If one function needs this requirement, even though there's a alternative method, then who's to say not every function could have this requirement? It's more a theoretical question as it is a practical one. But the point stands.

**Comment by GoranM** [comment]
> Most programmers would naturally assume that you don't have to add any objects to the scene in order for something like dimensionsLocal to work. It sure would be an exception to the rule. Which other method in BDX needs this? Because of that I also think it would be confusing. You would have to add notes about this in the API to counter this, I'm sure.

As things stand, you don't have access to template objects, so I don't see how you could ever get a reference to a GameObject that's not currently in the scene.

**Comment by rafcolson** [comment]
Well, this is all assuming that we could have access to template objects. A new ```dimensionsLocal``` (to replace the existing ```dimensions```) would be the only method that requires for objects to be added to the scene. I think such a feature is as relevant as ```scale()```, and the latter wouldn't have this requirement.

**Comment by GoranM** [comment]
I've given the go ahead for SolarLune to make templates public, so at this point, I think it would make sense to export dimensions, as you suggested.

However, I don't think there's a need for an additional function - We should simply change the implementation of the existing dimensions method.

**Comment by rafcolson** [comment]
Thanks, Goran. @SolarLune Let me know if I should assign myself to this. I'm willing to help.

**Comment by SolarLune** [comment]
Yeah, sure, feel free to work on it. Thanks!

**Comment by SolarLune** [comment]
Looks like (#320) solved this one. Thanks, @rafcolson! Issue closed.
